src/Raven.Server/Documents/Handlers/Processors/Batches/AbstractClusterTransactionRequestProcessor.cs

```diff
@@ -43,72 +46,88 @@ public async ValueTask<(long Index, DynamicJsonArray Results)> ProcessAsync(Json
+        ClusterTransactionCommand clusterTransactionCommand = CreateClusterTransactionCommand(parsedCommands, options, raftRequestId);
+        DynamicJsonArray array;
+        object result;
+        long index;
 
-            var clusterTransactionCommandResult = await RequestHandler.ServerStore.SendToLeaderAsync(clusterTransactionCommand);
-            if (clusterTransactionCommandResult.Result is List<ClusterTransactionCommand.ClusterTransactionErrorInfo> errors)
-            {
-                RequestHandler.HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
-                throw new ClusterTransactionConcurrencyException($"Failed to execute cluster transaction due to the following issues: {string.Join(Environment.NewLine, errors.Select(e => e.Message))}")
-                {
-                    ConcurrencyViolations = errors.Select(e => e.Violation).ToArray()
-                };
-            }
-            await RequestHandler.ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, clusterTransactionCommandResult.Index);
+        using (RequestHandler.ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(id: options.TaskId, out var onDatabaseCompletionTask))
+        {
+            (index, result) = await RequestHandler.ServerStore.SendToLeaderAsync(clusterTransactionCommand);
+            array = await GetClusterTransactionDatabaseCommandsResults(result, clusterTransactionCommand.DatabaseCommandsCount, index, onDatabaseCompletionTask);
+        }
 
-            DynamicJsonArray result;
-            if (clusterTransactionCommand.DatabaseCommands.Count > 0)
+        foreach (var clusterCommands in clusterTransactionCommand.ClusterCommands)
+        {
+            array.Add(new DynamicJsonValue
             {
-                using var cts = RequestHandler.CreateHttpRequestBoundTimeLimitedOperationToken(RequestHandler.ServerStore.Engine.OperationTimeout);
-                var databaseResult = await RequestHandler.ServerStore.Cluster.ClusterTransactionWaiter.WaitForResults(taskId, cts.Token);
+                [nameof(ICommandData.Type)] = clusterCommands.Type,
+                [nameof(ICompareExchangeValue.Key)] = clusterCommands.Id,
+                [nameof(ICompareExchangeValue.Index)] = index
+            });
+        }
 
-                if (databaseResult.IndexTask != null)
-                {
-                    await databaseResult.IndexTask;
-                }
+        return (index, array);
+    }
 
-                result = databaseResult.Array;
-            }
-            else
-            {
-                result = new DynamicJsonArray();
-            }
+    private async Task<DynamicJsonArray> GetClusterTransactionDatabaseCommandsResults(object result, long databaseCommandsCount, long index, Task<Task> onDatabaseCompletionTask)
+    {
+        // old leader returned errors
```

Comment:
note

Reply:
Deleted, This note isn't relevant anymore..